llvm/lib/Target/AIE/AIEBaseAliasAnalysis.cpp

```diff
@@ -117,13 +153,300 @@ static const Value *getUnderlyingObjectAIE(const Value *V) {
+    // This information is crucial to calculate the number
+    // of pointer updates before TargetPointer.
+    if (LastUpdate == TargetPointer)
+      LastCalls = ChainInfo.TotalIntrinsicCalls;
+    assert(LastUpdate->getType()->isPointerTy() && "Unexpected operand type!");
+  }
+  return std::nullopt;
+}
+
+static AliasResult aliasAIEIntrinsic(const Value *ValueA, const Value *ValueB,
+                                     const Value *BaseA, const Value *BaseB,
+                                     const unsigned VirtUnrollLevelA,
+                                     const unsigned VirtUnrollLevelB) {
+
+  const PHINode *PhiA = dyn_cast<PHINode>(BaseA);
+  const PHINode *PhiB = dyn_cast<PHINode>(BaseB);
+
+  // Both bases must be PHI nodes.
+  if (!PhiA || !PhiB)
+    return AliasResult::MayAlias;
+
+  // Skip non-pointer values (cast from pointers?).
+  if (!PhiA->getType()->isPointerTy() || !PhiB->getType()->isPointerTy())
+    return AliasResult::MayAlias;
+
+  // Our target is loop-recurrent PHI nodes.
+  if (PhiA->getNumIncomingValues() != 2 || PhiB->getNumIncomingValues() != 2)
+    return AliasResult::MayAlias;
+
+  // Both PHI nodes must be in the same MBB.
+  if (PhiA->getParent() != PhiB->getParent())
+    return AliasResult::MayAlias;
+
+  auto GetExternalIndex = [](const PHINode *Phi) {
+    return Phi->getIncomingBlock(0) == Phi->getParent() ? 1 : 0;
+  };
+
+  auto GetInternalIndex = [](const PHINode *Phi) {
+    return Phi->getIncomingBlock(0) == Phi->getParent() ? 0 : 1;
+  };
+
+  const Value *ExtPhiValA = PhiA->getIncomingValue(GetExternalIndex(PhiA));
+  const Value *ExtPhiValB = PhiB->getIncomingValue(GetExternalIndex(PhiB));
+
+  // Both Phis must have the same value as an external incoming operand,
+  // so we guarantee that we have two new pointers rooted in the same
+  // original pointer.
+  //
+  // NOTE: We don't care if PhiA and PhiB are the same or not, so we can also
+  // disambiguate different steps starting from from the same pointer (Phi).
+  // We only care if the external value is the same.
+  // For example, we can have the following case:
+  // a =  phi ptr [ %an, %this_mbb ], [ %a_ext, %pred_mbb ]
+  // x1 = load (a)
+  // a1 = add_2d(a, /*fixed parameters + chained counters*/)
+  // x2 = load (a1)
+  // an = add_2d(a1, /*fixed parameters + chained counters*/)
+  // Where we can prove that those two loads are not aliasing, even though
+  // they are sharing the same phi node as an underlying object.
+  if (ExtPhiValA != ExtPhiValB)
+    return AliasResult::MayAlias;
+
+  // Most of the remaining code aims to prove that the two pointers
+  // are updated with same steps.
+
+  const Value *IntPhiValA = PhiA->getIncomingValue(GetInternalIndex(PhiA));
+  const Value *IntPhiValB = PhiB->getIncomingValue(GetInternalIndex(PhiB));
+
+  // Now we know that the pointers are updated consistently across
+  // loop iterations, let's look to the values.
+  auto ATracking = trackIntrinsicChain(IntPhiValA, ValueA);
```

Comment:
Hi @gbossu, another simplification round, according to your comments.

Now I have just 2 calls to `trackIntrinsicChain`. I already need to have the phi nodes and the incoming values. The external value is used to fast skip some obvious cases. 

In this way, we track all updates between the "Internal value" to the direction of the phi node. During this traversal, we will find the pointer of interest and get some information to precisely calculate the number of updated between phi and this node. We need to start from the last update (internal value) to the phi node to know precisely the number of updates across the loop (this information is also required to virtual unroll).

Also, all verifications are encapsulated in `IntrinsicChainInfo` except the initial phi validation. 

Thank you for all suggestions so far.